tapgarden/planter.go

```diff
@@ -724,10 +725,134 @@ func freezeMintingBatch(ctx context.Context, batchStore MintingStore,
 	)
 }
 
+// filterFinalizedBatches separates a set of batches into two sets based on
+// their batch state.
+func filterFinalizedBatches(batches []*MintingBatch) ([]*MintingBatch,
+	[]*MintingBatch) {
+
+	finalized := []*MintingBatch{}
+	nonFinalized := []*MintingBatch{}
+
+	fn.ForEach(batches, func(batch *MintingBatch) {
+		switch batch.State() {
+		case BatchStateFinalized:
+			finalized = append(finalized, batch)
+		default:
+			nonFinalized = append(nonFinalized, batch)
+		}
+	})
+
+	return finalized, nonFinalized
+}
+
+// fetchFinalizedBatch fetches the assets of a batch in their genesis state,
+// given a batch populated with seedlings.
+func fetchFinalizedBatch(ctx context.Context, archiver proof.Archiver,
+	batch *MintingBatch) (*MintingBatch, error) {
+
+	// Collect genesis TX information from the batch to build the proof
+	// locators.
+	anchorOutputIndex := uint32(0)
+	if batch.GenesisPacket.ChangeOutputIndex == 0 {
+		anchorOutputIndex = 1
+	}
+	signedTx, err := psbt.Extract(batch.GenesisPacket.Pkt)
+	if err != nil {
+		return nil, err
+	}
+
+	genOutpoint := extractGenesisOutpoint(signedTx)
+	genScript := signedTx.TxOut[anchorOutputIndex].PkScript
+	anchorOutpoint := wire.OutPoint{
+		Hash:  signedTx.TxHash(),
+		Index: anchorOutputIndex,
+	}
+
+	batchAssets := make([]*asset.Asset, 0, len(batch.Seedlings))
+	assetMetas := make(AssetMetas)
+	for _, seedling := range batch.Seedlings {
+		gen := seedling.Genesis(genOutpoint, anchorOutputIndex)
+		scriptKey := asset.ToSerialized(seedling.ScriptKey.PubKey)
+		assetMetas[scriptKey] = seedling.Meta
+
+		assetLoc := proof.Locator{
+			AssetID:   fn.Ptr(gen.ID()),
```

Comment:
Missing `GroupKey` here. WTB #865

Reply:
Though we if _know_ that this is the file based archive, then things are fine here.

Reply:
Yup, `b.cfg.ProofFiles` only requires a `proof.Archiver` implementation, but in practice its always the `FileArchiver`.